src/apps/vpn/ui/screens/settings/ViewDNSSettings.qml

```diff
@@ -180,21 +180,16 @@ VPNViewBase {
                         ipInput.text = VPNSettings.userDNS;
                     }
 
-                    onTextChanged: text => {
-                        if (ipInput.text === "") {
-                            // If nothing is entered, thats valid too. We will ignore the value later.
-                            ipInput.valueInvalid = false;
+                    onEditingFinished: {
+                        if (ipInput.text === "" || VPN.validateUserDNS(ipInput.text)) {
                             VPNSettings.userDNS = ipInput.text
-                            return;
-                        }
-                        if (VPN.validateUserDNS(ipInput.text)) {
-                            ipInput.valueInvalid = false;
-                            VPNSettings.userDNS = ipInput.text
-                        } else {
-                            ipInput.error = VPNI18n.SettingsDnsSettingsCustomDNSError
-                            ipInput.valueInvalid = true;
                         }
                     }
+
```

Comment:
Incidentally, you'll have fixed most of [VPN-4070](https://mozilla-hub.atlassian.net/browse/VPN-4070) if you add something like the following to this VPNTextField.

```
  Component.onDestruction: {
      if ((VPNSettings.dnsProviderFlags === VPNSettings.Custom) && ipInput.error) {
          maybeApplyChange(VPNSettings.Gateway);
      }
  }
```